lib/awful/screen/dpi.lua

```diff
@@ -0,0 +1,414 @@
+---------------------------------------------------------------------------
+--                          DPI detection code.                          --
+---------------------------------------------------------------------------
+
+local capi    = {screen = screen}
+local gtable  = require("gears.table")
+local grect   = require("gears.geometry").rectangle
+local gdebug  = require("gears.debug")
+
+local module = {}
+
+local ascreen, data = nil, nil
+
+-- Metric to Imperial conversion constant.
+local mm_per_inch = 25.4
+
+local xft_dpi, fallback_dpi
+
+local function get_fallback_dpi()
+    -- Following Keith Packard's whitepaper on Xft,
+    -- https://keithp.com/~keithp/talks/xtc2001/paper/xft.html#sec-editing
+    -- the proper fallback for Xft.dpi is the vertical DPI reported by
+    -- the X server. This will generally be 96 on Xorg, unless the user
+    -- has configured it differently
+    if not fallback_dpi then
+        local _, h = root.size()
+        local _, hmm = root.size_mm()
+        fallback_dpi = hmm ~= 0 and h * mm_per_inch / hmm
+    end
+
+    return math.floor(fallback_dpi or 96)
+end
+
+local function dpi_for_output(viewport, output)
+    local dpi = nil
+    local geo = viewport.geometry
+
+    -- Ignore outputs with width/height 0
+    if output.mm_width ~= 0 and output.mm_height ~= 0 then
+        local dpix = geo.width * mm_per_inch / output.mm_width
+        local dpiy = geo.height * mm_per_inch / output.mm_height
+        dpi = math.min(dpix, dpiy, dpi or dpix)
+    elseif ascreen._get_xft_dpi() then
+        dpi = ascreen._get_xft_dpi()
```

Comment:
This can be simplified to `else dpi = ascreen._get_xft_dpi()`. The function is called one time less and the result is only used in an `dpi or foo`, so the result should be equivalent.